src/object/srv_ec_aggregate.c

```diff
@@ -1587,23 +1601,33 @@ agg_process_holes(struct ec_agg_entry *entry)
 		rc = dss_abterr2der(rc);
 		goto ev_out;
 	}
-
+#if HUH
+	if (*status != 0) {
+		rc = *status;
+		goto ev_out;
+	}
+#endif
 	/* Update local vos with replicate */
 	iod.iod_name = entry->ae_akey;
 	iod.iod_type = DAOS_IOD_ARRAY;
 	iod.iod_size = entry->ae_rsize;
 	iod.iod_nr = stripe_ud.asu_cell_cnt;
 	iod.iod_recxs = stripe_ud.asu_recxs;
+
 	entry->ae_sgl.sg_nr = 1;
-	/* write the reps to vos */
-	agg_param = container_of(entry, struct ec_agg_param, ap_agg_entry);
-	rc = vos_obj_update(agg_param->ap_cont_handle, entry->ae_oid,
-			    entry->ae_cur_stripe.as_hi_epoch, 0, 0,
-			    &entry->ae_dkey, 1, &iod, NULL,
-			    &entry->ae_sgl);
-	if (rc) {
-		D_ERROR("vos_update_begin failed: "DF_RC"\n", DP_RC(rc));
-		goto ev_out;
+	agg_param = container_of(entry, struct ec_agg_param,
+				 ap_agg_entry);
+	if (iod.iod_nr) {
+		/* write the reps to vos */
+		rc = vos_obj_update(agg_param->ap_cont_handle, entry->ae_oid,
+				    entry->ae_cur_stripe.as_hi_epoch, 0, 0,
+				    &entry->ae_dkey, 1, &iod, NULL,
+				    &entry->ae_sgl);
+		if (rc) {
```

Comment:
hmm, it create an ULT above, 

	tid = dss_get_module_info()->dmi_tgt_id;
	rc = dss_ult_create(agg_process_holes_ult, &stripe_ud,
			    DSS_XS_OFFLOAD, tid, 0, NULL);
	if (rc)

So this process might not be run on the VOS target ULT, why it is using vos_obj_update() directly here?  otherwise, why create an ULT above?

Reply:
The call to vos_obj_update (the lines you're commenting on) runs on the VOS target ULT.  The offload is for the internode communication (remote fetch of replicas and update of peer parity nodes).  The code is offloaded because it does not access local VOS and should not block the Xstream while doing RPCs.

Reply:
Ah, agg_process_holes_ult() will not run on the local vos target ult. I see. Thanks.